core/src/main/kotlin/rys/ajaxpetproject/service/impl/ChatServiceImpl.kt

```diff
@@ -1,36 +1,122 @@
+    private val chatRepository: ChatRepository,
+    private val userService: UserService,
+    private val messageService: MessageService
+) : ChatService {
+    override fun findChatById(id: String): Mono<MongoChat> {
+        return chatRepository.findChatById(id)
+    }
+
+    override fun save(chat: MongoChat): Mono<MongoChat> {
+        return chatRepository.save(chat)
+    }
+
+    override fun deleteAll(): Mono<Unit> {
+        return chatRepository.deleteAll()
+    }
 
-    override fun createChat(mongoChat: MongoChat): MongoChat {
-        return chatRepository.save(mongoChat)
+    override fun update(id: String, chat: MongoChat): Mono<MongoChat> {
+        return findChatById(id)
+            .switchIfEmpty { Mono.error(ChatNotFoundException("Chat with id $id not found")) }
+            .flatMap { chatRepository.update(id, chat) }
     }
 
-    override fun findChatById(id: ObjectId): MongoChat? = chatRepository.findChatById(id)
+    override fun addUser(userId: String, chatId: String): Mono<Unit> {
+        return Mono.`when`(
+            userService.findUserById(userId)
+                .switchIfEmpty { Mono.error(UserNotFoundException("User with id $userId not found")) },
+            findChatById(chatId)
+                .switchIfEmpty { Mono.error(ChatNotFoundException("Chat with id $chatId not found")) }
+        )
+            .then(chatRepository.addUser(userId, chatId))
+    }
 
-    override fun findAllChats(): List<MongoChat> = chatRepository.findAllBy()
+    override fun removeUser(userId: String, chatId: String): Mono<Unit> {
+        return Mono.`when`(
+            userService.findUserById(userId)
+                .switchIfEmpty { Mono.error(UserNotFoundException("User with id $userId not found")) },
+            findChatById(chatId)
+                .switchIfEmpty { Mono.error(ChatNotFoundException("Chat with id $chatId not found")) }
+        )
+            .then(chatRepository.removeUser(userId, chatId))
+    }
+
+    override fun addMessage(messageId: String, chatId: String): Mono<Unit> {
+        return Mono.`when`(
+            findChatById(chatId)
+                .switchIfEmpty { Mono.error(ChatNotFoundException("Chat with id $chatId not found")) },
+            messageService.findMessageById(messageId)
+                .switchIfEmpty { Mono.error(ChatNotFoundException("Message with id $messageId not found")) }
+        )
+            .then(chatRepository.addMessage(messageId, chatId))
+    }
 
-    override fun updateChat(id: ObjectId, updatedMongoChat: MongoChat): MongoChat {
+    override fun removeMessage(messageId: String, chatId: String): Mono<Unit> {
+        return chatRepository.removeMessage(messageId, chatId)
+    }
+
+    override fun delete(id: String): Mono<Unit> {
         return findChatById(id)
-            .let {
-                chatRepository.save(updatedMongoChat.copy(id = id))
-            }
+            .switchIfEmpty { Mono.error(ChatNotFoundException("Chat with id $id not found")) }
```

Comment:
Find chat by id + throw if chat not found is repeated 8 times in this class

![image](https://github.com/BohdanRysakov/AjaxOnBoardRysakov/assets/116563306/1428ba15-56ab-4b2c-adf0-c3ce0aa82367)

Reply:
Just informing you that you have multiple duplicated code.

` getChatById()` - good solution

Reply:
Done